<a href="https://colab.research.google.com/github/nurulshifaside/Data-Science-Tasks/blob/main/4th%20Task/Introduction%20to%20Artificial%20Intelligence.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Assignment 5 - Introduction to Artificial Intelligence

**Team 7A**


Anggota:
1. Muhammad Fikri Abdillah Syaeba
2. Agnes Theresia Siburian
3. Raphael Junus Samson
4. Nurul Shifa Maulina
5. Angel Paulina Siahaan
6. Octavia Kustiana
7. Hanisa Feranti

## Question

**2. Implementation of Artificial Intelligence**

Text classification berdasarkan Bag-of-Words

Tulis syntax di Google Colab:
- Tentukan tiga topik terkait: Kesehatan, Olahraga, Keuangan.
- Temukan teks/artikel sebagai dasar/pelatihan/training ketiga topik tersebut.
- Hitung vektor Bag-of-Words untuk setiap topik.

Klasifikasikan teks ini dalam topik yang benar:
“Cristiano Ronaldo came off the bench to earn Manchester United a hard-fought 2-1 victory at Everton in the Premier League on Sunday, taking his career goal tally to 700 in the process. Just as United did last weekend in their derby mauling at the hands of local rivals Manchester City, they again found themselves behind early on at Goodison Park after Alex Iwobi curled a sublime strike into the net from 20 metres.”

###**1. Mengimpor Library yang Diperlukan**

In [ ]:
from nltk.tokenize import word_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

In [ ]:
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.metrics.pairwise import cosine_similarity
import pandas as pd
import numpy as np

In [ ]:
import nltk
nltk.download("popular")
from nltk.tokenize import sent_tokenize, word_tokenize

[nltk_data] Downloading collection 'popular'
[nltk_data]    | 
[nltk_data]    | Downloading package cmudict to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/cmudict.zip.
[nltk_data]    | Downloading package gazetteers to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/gazetteers.zip.
[nltk_data]    | Downloading package genesis to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/genesis.zip.
[nltk_data]    | Downloading package gutenberg to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/gutenberg.zip.
[nltk_data]    | Downloading package inaugural to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/inaugural.zip.
[nltk_data]    | Downloading package movie_reviews to
[nltk_data]    |     /root/nltk_data...
[nltk_data]    |   Unzipping corpora/movie_reviews.zip.
[nltk_data]    | Downloading package names to /root/nltk_data...
[nltk_data]    |   Unzipping corpora/names.zip.
[nltk_data]    | Downloading package shakespeare to /root/nltk_data...
[nlt

###**2. Membaca Dataframe yang Telah Dibuat**

Membaca dataframe yang berisi teks-teks untuk setiap topik.

In [ ]:
df_team_7A = pd.read_csv('https://docs.google.com/spreadsheets/d/e/2PACX-1vSd3tKknOavUlxa2YC8qmigvvaFpMvmGs9jxGyInhqUYE8ETJ4dWfJR8CTT4KPguB2pOoKkB-ZtZ21c/pub?gid=0&single=true&output=csv')
df_team_7A

Text      Topic     Name
0  In today's rapidly evolving financial landscap...   Keuangan    Fikri
1  The associations between short or long sleep d...  Kesehatan  Raphael
2  The final Merseyside derby of Jürgen Klopp’s L...   Olahraga  Raphael
3  The village head of her community told Baitul ...   Keuangan    Nurul
4  The Dangers of Smoking\n\nSmoking remains a le...  Kesehatan   Viiaaa
5  Real Madrid defeats Manchester City in penalty...   Olahraga    Agnes
6  Who said you cannot do an exercise while sitti...   Olahraga    Angel

###**3. Menggabungkan Teks dari Dataframe ke dalam List**

Menggabungkan teks-teks dalam masing-masing topik ke dalam satu list gabungan.

In [ ]:
# Make a list of each text topic
texts_sport = df_team_7A[df_team_7A['Topic'] == 'Olahraga']['Text'].tolist()
texts_medical = df_team_7A[df_team_7A['Topic'] == 'Kesehatan']['Text'].tolist()
texts_finance = df_team_7A[df_team_7A['Topic'] == 'Keuangan']['Text'].tolist()

# Combine texts from the same topic
combined_texts_sport = ' '.join(texts_sport)
combined_texts_medical = ' '.join(texts_medical)
combined_texts_finance = ' '.join(texts_finance)

# Combines all text in one list
all_texts = []
all_texts.extend([combined_texts_sport])
all_texts.extend([combined_texts_medical])
all_texts.extend([combined_texts_finance])

###**4. Membersihkan Teks**

Melakukan pembersihan teks dengan menghapus stopwords dan mengaplikasikan lemmatization.

In [ ]:
# Cleaning Teks
lemmatizer = WordNetLemmatizer() # Create object for lemmatizer
bow = [] # Bag of Word
kumpulan_text = []
for text in all_texts:
    words  = word_tokenize(text)
    all_texts = [lemmatizer.lemmatize(word) for word in words if word not in set(stopwords.words('english'))]
    bow += all_texts
    text = ' '.join(all_texts)
    kumpulan_text.append(text)
bow = set(bow)
print(bow)
print(kumpulan_text)

{'inconsequential', 'organ', ';', 'League', 'ended', 'defense', 'Park', 'compared', 'penalty', 'well-being', 'thus', ',', 'crowd', 'sitting', 'association', 'filled', 'allows', 'One', 'scored', 'adverse', 'take', 'track', '2022', 'rapidly', 'continues', 'offer', '4-3', 'team', '>', 'chronic', 'worker', 'Since', 'reverse', 'worldwide', 'Jürgen', '7–8', 'marker', 'way', 'spending', 'tobacco', 'certain', 'Real', 'tasted', 'benefiting', 'coronary', 'well-documented', 'remains', 'information', 'thrive', 'Victory', 'back', 'bellowed', 'enjoy', 'outcome', 'community', 'manage', 'help', 'income', 'remembered', 'prevent', 'defeat', 'reign', 'league', 'much', 'semifinal', 'year', 'almost', 'embrace', 'programme', 'might', 'high', 'From', 'work', 'certainly', 'treble', 'known', 'respectively', 'control', 'lung', 'Exercise', 'budget', 'need', 'thrilling', 'informed', 'effectively', 'Officer', 'home', 'time', 'saw', 'Jarrad', 'City', 'Moreover', 'habit', 'budgeting', 'Manchester', 'tie', 'blow', 'm

###**5. Memasukkan, Membersihkan, dan Membuat Vektor pada Teks Baru**

Membuat vektor Bag-of-Words untuk setiap teks, termasuk teks baru yang akan diklasifikasikan.

In [ ]:
# A new text to be classified based on topic
query_text = "Cristiano Ronaldo came off the bench to earn Manchester United a hard-fought 2-1 victory at Everton in the Premier League on Sunday, taking his career goal tally to 700 in the process. Just as United did last weekend in their derby mauling at the hands of local rivals Manchester City, they again found themselves behind early on at Goodison Park after Alex Iwobi curled a sublime strike into the net from 20 metres."
query_words = word_tokenize(query_text)
query_words_clean = [lemmatizer.lemmatize(word) for word in query_words if word not in set(stopwords.words('english'))]
query_kumpulan_kata = [word for word in query_words_clean if word in set(bow)]
query_kumpulan_text = ' '.join(query_kumpulan_kata)
kumpulan_text.append(query_kumpulan_text)

cv = CountVectorizer() # Creating Object for CountVectorizer
bow_vectors = cv.fit_transform(kumpulan_text).toarray()
print(bow_vectors)
print(len(bow_vectors[0]))

[[0 1 1 ... 0 0 0]
 [0 0 0 ... 0 0 0]
 [1 0 0 ... 1 1 1]
 [0 0 0 ... 0 0 0]]
395


###**6. Menghitung Persentase Similarity**

Menghitung persentase kesamaan antara vektor teks baru dengan vektor topik yang sudah ada.

In [ ]:
# Normalize the BoW vectors
bow_texts_norm = []
for x in bow_vectors:
  length = (sum(i*i for i in x)) ** 0.5
  bow_norm = x / length
  bow_texts_norm.append(bow_norm)

# Compute similarity using dot product
similarity_vector = []
bow_norm_query = bow_texts_norm[3]
for x in bow_texts_norm[:3]:
  similarity_vector.append(sum(i*j for i,j in zip(x,bow_norm_query)))

# Calculate the percentage of each topic
sport_similarity_percentage = (similarity_vector[0]) * 100
medical_similarity_percentage = (similarity_vector[1]) * 100
finance_similarity_percentage = (similarity_vector[2]) * 100

# Prints the percentage of each topic with two decimal places after the comma
print("Persentase kesamaan dengan topik olahraga: {:.2f}%".format(sport_similarity_percentage))
print("Persentase kesamaan dengan topik kesehatan: {:.2f}%".format(medical_similarity_percentage))
print("Persentase kesamaan dengan topik keuangan: {:.2f}%".format(finance_similarity_percentage))

Persentase kesamaan dengan topik olahraga: 34.65%
Persentase kesamaan dengan topik kesehatan: 2.27%
Persentase kesamaan dengan topik keuangan: 2.82%


### **7. Menentukan Topik Teks**


Menentukan topik yang memiliki persentase kesamaan tertinggi sebagai klasifikasi teks.

In [ ]:
# Find the highest similarity
max_similatity = similarity_vector.index(max(similarity_vector))
if (max_similatity == 0):
  print ("Teks Diklasifikasikan sebagai Teks Olahraga")
elif (max_similatity == 1):
  print ("Teks Diklasifikasikan sebagai Teks Kesehatan")
elif (max_similatity == 2):
  print ("Teks Diklasifikasikan sebagai Teks Keuangan")

Teks Diklasifikasikan sebagai Teks Olahraga


 Teks yang diberikan berhasil diklasifikasikan sebagai **Teks Olahraga** dengan persentase kesamaan sebesar **34.65%**. Persentase kesamaan dengan topik kesehatan dan keuangan jauh lebih rendah, yaitu **2.27%** dan **2.82%** secara berturut-turut. Oleh karena itu, kesimpulan dari hasil implementasi ini adalah bahwa teks tersebut dapat diklasifikasikan sebagai teks olahraga berdasarkan model Bag-of-Words yang telah dibangun.